### ISSUE TITLE
Program casues random black screen 

### ISSUE THREAD

**Issue by stblackburn** [Rem0o/FanControl.Releases#302]
Just found this app and im impressed however it seems as if almost the GPU crashes and the screen goes black and causes the whole PC to start to slow down or weird graphical issues. No messages from the GPU or system showing a grahpical driver crash however event veiwer shows the following: 

**Error 1:** 
Faulting application name: FanControl.exe, version: 1.0.0.0, time stamp: 0xb9a6e448
Faulting module name: unknown, version: 0.0.0.0, time stamp: 0x00000000
Exception code: 0xc0000005
Fault offset: 0x00007ffdeb39b94f
Faulting process id: 0x8248
Faulting application start time: 0x01d780d293384d8b
Faulting application path: ------\Downloads\FanControl.Releases-87\FanControl.Releases-87\FanControl\FanControl.exe
Faulting module path: unknown
Report Id: eafc3081-61d6-4d2e-ad1a-0df23d729f2e
Faulting package full name: 
Faulting package-relative application ID: 

**Error 2:** 

Application: FanControl.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.NullReferenceException
   at FanControl.Domain.BackendProviders.LHM.LHMBackendProvider.Close()
   at System.Collections.Generic.List`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].ForEach(System.Action`1<System.__Canon>)
   at FanControl.Domain.ComputerAccessLayer.Close()
   at FanControl.Domain.ComputerAccessLayer.Dispose()
   at Autofac.Core.Activators.ProvidedInstance.ProvidedInstanceActivator.Dispose(Boolean)
   at Autofac.Util.Disposable.Dispose()
   at Autofac.Core.Registration.ComponentRegistration.Dispose(Boolean)
   at Autofac.Util.Disposable.Dispose()
   at Autofac.Core.Registration.DefaultRegisteredServicesTracker.Dispose(Boolean)
   at Autofac.Util.Disposable.Dispose()
   at Autofac.Core.Registration.ComponentRegistry.Dispose(Boolean)
   at Autofac.Util.Disposable.Dispose()
   at Autofac.Core.Container.Dispose(Boolean)
   at Autofac.Util.Disposable.Dispose()
   at FanControl.App.OnExit(System.Windows.ExitEventArgs)
   at System.Windows.Application.DoShutdown()
   at System.Windows.Application.ShutdownImpl()
   at System.Windows.Application.ShutdownCallback(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
   at MS.Internal.CulturePreservingExecutionContext.CallbackWrapper(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at MS.Internal.CulturePreservingExecutionContext.Run(MS.Internal.CulturePreservingExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Windows.Threading.DispatcherOperation.Invoke()
   at System.Windows.Threading.Dispatcher.ProcessQueue()
   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndWrapper.WndProc(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority, System.TimeSpan, System.Delegate, System.Object, Int32)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr, Int32, IntPtr, IntPtr)
   at MS.Win32.UnsafeNativeMethods.DispatchMessage(System.Windows.Interop.MSG ByRef)
   at System.Windows.Threading.Dispatcher.PushFrameImpl(System.Windows.Threading.DispatcherFrame)
   at System.Windows.Application.RunDispatcher(System.Object)
   at System.Windows.Application.RunInternal(System.Windows.Window)
   at FanControl.App.Main()

**WIndows Error:** 

Fault bucket 1994371379912959065, type 5
Event Name: CLR20r3
Response: Not available
Cab Id: 0

Problem signature:
P1: FanControl.exe
P2: 1.0.0.0
P3: b9a6e448
P4: FanControl
P5: 1.0.0.0
P6: b9a6e448
P7: 5b8
P8: 5c
P9: System.NullReferenceException
P10: 

Attached files:
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER50BF.tmp.dmp
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER513D.tmp.WERInternalMetadata.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER514E.tmp.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER514E.tmp.csv
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER516E.tmp.txt

These files may be available here:
\\?\C:\ProgramData\Microsoft\Windows\WER\ReportArchive\AppCrash_FanControl.exe_fbbb9d42c9efd6365e312e2b5821f55b0478ac4_1002a926_2b897139-8c4c-44c1-84d9-7b44576d34b2

Analysis symbol: 
Rechecking for solution: 0
Report Id: eafc3081-61d6-4d2e-ad1a-0df23d729f2e
Report Status: 268435456
Hashed bucket: 686eca695bbe62da4bad6e3411eae459
Cab Guid: 0

**Comment by abqnm** [comment]
Getting the same issue. It is just slow to run and then the screen goes black and the graphics driver crashes. Has happened on both Nvidia 471.68 & 471.41, haven't tried an older version. Only happens when the UI is visible, it functions normally when the UI is minimized. 

But it's becoming more and more frequent, now happening nearly anytime I open the main interface.

Of note, the window also flickers constantly when open. Not sure if that has something to do with it (is it re-drawing the entire interface multiple times per second or something?).

Ryzen 5800x, Gigabyte x570-I, RTX 3070 running stock but with reduced power limit, no change at 100% power limit.

**Comment by abqnm** [comment]
[FanControl-Output-Debug.txt](https://github.com/Rem0o/FanControl.Releases/files/7003440/FanControl-Output-Debug.txt)

Here is a debug log from Visual Studio.  The majority of the exceptions occur constantly in the background (I have never run the debugger on it before so I don't know if those background exceptions were always happening or not), but the graphics driver crash occurs exactly at these two lines (lines 744 & 745) which say the following:

'FanControl.exe' (Win32): Loaded 'C:\Windows\System32\DriverStore\FileRepository\nv_dispi.inf_amd64_253e24b785ea60ca\NvCamera\NvCameraAllowlisting64.dll'. 
'FanControl.exe' (Win32): Unloaded 'C:\Windows\System32\DriverStore\FileRepository\nv_dispi.inf_amd64_253e24b785ea60ca\NvCamera\NvCameraAllowlisting64.dll'

Hopefully this can help you find the issue, as it is quite significant to have a 2D application crashing the GPU driver just by simply being visible on the display.

**Comment by Rem0o** [comment]
"Exception thrown at 0x00007FFF11724ED9 in FanControl.exe: Microsoft C++ exception: FileFormatException at memory location 0x000000BE5813E3F8."

Seems like it is trying to load a 64 bit dll and FanControl is a 32 bit application, which explains the "FileFormatException".

The problem is, I have no idea what NvCameraAllowlisting64 is, or why/how it has anything to my application UI, which is a really basic WPF/xaml app.

The Nvidia driver though is accessed with https://github.com/falahati/NvAPIWrapper to get temperature and fan/rpm reading. That's the closest thing I can think of.

Which version of Windows btw?

**Comment by abqnm** [comment]
Yeah it's hard to say if it's the GUI causing it, but that's the only time the crash occurs is when the GUI is visible. But I think you're right that it probably has to do with the NvAPIWrapper, as it's definitely an Nvidia crash.

I've also done some more probing in the last 2 days, and found that aside from the first time when I rolled back directly from 471.68, I can't get it to happen on the Nvidia 471.41 driver after using DDU. I had previously just rolled back in place, and the crashes continued. But now that I rolled back by using DDU to remove the drivers completely then reinstalled 471.41 from scratch, it hasn't crashed since. 

So I'm thinking it may be the 471.68 drivers changed something that NvAPIWrapper is calling, as you said since that's a 32-

W10 21H1, (build 19043.1200).

**Comment by PeterPan669** [comment]
Did you ever solve this issue? I noticed that I got the same behavior (black screen, seems like power loss in some form as keyboard lights are off as well), this started happening when I set 'User Fan Control + Firmware Control' in MSI Afterburner. It became more frequent, even happening at a simple Teams meeting when there was not much load at all on the system. After disabling those settings, I started getting 'normal' hangups in RDR2. Now, those hangups can be caused by RDR2 as the game seems buggy as hell, but the whole black screen thing happened same way you describe. P.S., the event log errors you see regarding the application is related to the whole system crashing / going down, and the ETW writing the error for the app in the  log. The real culprit doesn't need to be Fan Control, it's simply reporting a crash to the ETW.
